core/src/com/google/inject/internal/BytecodeGen.java

```diff
@@ -16,412 +16,147 @@
-        }
-      };
-
-  static final net.sf.cglib.core.NamingPolicy ENHANCER_NAMING_POLICY =
-      new net.sf.cglib.core.DefaultNamingPolicy() {
-        @Override
-        protected String getTag() {
-          return "ByGuice";
-        }
-
-        @Override
-        public String getClassName(
-            String prefix, String source, Object key, net.sf.cglib.core.Predicate names) {
-          // we explicitly set the source here to "Enhancer" so that our jarjar renaming
-          // to $Enhancer doesn't leak into the class names.  if we did not do this,
-          // classes would end up looking like $$$EnhancerByGuice$$, with the extra $
-          // at the front.
-          return super.getClassName(prefix, "Enhancer", key, names);
-        }
-      };
-  /*end[AOP]*/
-  /*if[NO_AOP]
-  private static final String CGLIB_PACKAGE = " "; // any string that's illegal in a package name
-  end[NO_AOP]*/
 
-  /**
-   * Weak cache of bridge class loaders that make the Guice implementation classes visible to
-   * various code-generated proxies of client classes.
-   */
-  private static final LoadingCache<ClassLoader, ClassLoader> CLASS_LOADER_CACHE;
+  public static final String ENHANCER_BY_GUICE_MARKER = "$$EnhancerByGuice$$";
 
-  static {
-    CacheBuilder<Object, Object> builder = CacheBuilder.newBuilder().weakKeys().weakValues();
-    if (getCustomClassLoadingOption() == CustomClassLoadingOption.OFF) {
-      builder.maximumSize(0);
-    }
-    CLASS_LOADER_CACHE =
-        builder.build(
-            new CacheLoader<ClassLoader, ClassLoader>() {
-              @Override
-              public ClassLoader load(final ClassLoader typeClassLoader) {
-                logger.fine("Creating a bridge ClassLoader for " + typeClassLoader);
-                return AccessController.doPrivileged(
-                    new PrivilegedAction<ClassLoader>() {
-                      @Override
-                      public ClassLoader run() {
-                        return new BridgeClassLoader(typeClassLoader);
-                      }
-                    });
-              }
-            });
-  }
+  public static final String FASTCLASS_BY_GUICE_MARKER = "$$FastClassByGuice$$";
 
-  /**
-   * Attempts to canonicalize null references to the system class loader. May return null if for
-   * some reason the system loader is unavailable.
-   */
-  private static ClassLoader canonicalize(ClassLoader classLoader) {
-    return classLoader != null ? classLoader : SystemBridgeHolder.SYSTEM_BRIDGE.getParent();
-  }
+  /** Builder of enhanced classes. */
+  public interface EnhancerBuilder {
+    /** Lists the methods in the host class that can be enhanced. */
+    Method[] getEnhanceableMethods();
 
-  /** Returns the class loader to host generated classes for {@code type}. */
-  public static ClassLoader getClassLoader(Class<?> type) {
-    return getClassLoader(type, type.getClassLoader());
+    /** Generates an enhancer for the selected subset of methods. */
```

Comment:
I'll work on documenting this further - note that using `Function` and `BiFunction` is a particular point of the new design because it means the generated code is completely decoupled from Guice types, which makes loading it into the target class loader a lot easier. If we used a special type here then the generated class would need access to it in order to implement it, which means using a bridge class loader in modular environments which gets messy.

Reply:
It would be good to also document why `Function` and `BiFunction` is used here.

Reply:
f1129c3